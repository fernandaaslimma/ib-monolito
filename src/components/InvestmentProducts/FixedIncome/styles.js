import styled from "styled-components";
import { darkGreen, white, lighestgrey } from "../../../styles/settings";
import { rem, remFontSize } from "../../../styles/tools";

export const Title = styled.h1`
  font-size: ${remFontSize(20)};
  padding: ${rem(25)} 0;
  font-family: "Lato Bold", Lato;
  color: ${darkGreen};
`;

export const StickyWrapper = styled.div`
  overflow: auto;
  position: sticky;
  bottom: 0;
  background: ${white};
  display: flex;
`;

export const BtnWrapper = styled.div`
  flex-grow: 1;
  position: relative;
  flex-basis: 0;
  display: flex;

  > div {
    flex-grow: 1;
    flex-basis: 0;
  }
`;

export const Separator = styled.span`
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: ${rem(1)};
  background: ${lighestgrey};
`;

export const EFTTokenMfaWrapper = styled.div`
  margin: auto auto ${rem(16)} auto;
`;

export const StepVisibility = styled.span`
  height: 0px;
`;

export const ClickableItem = styled.span`
  cursor: pointer;
`;
