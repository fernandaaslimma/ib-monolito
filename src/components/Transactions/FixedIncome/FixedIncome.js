import React, { Component } from "react";
import {
  string,
  arrayOf,
  shape,
  number,
  func,
  oneOfType,
  instanceOf
} from "prop-types";

import Hide from "../../common/Hide";
import Mobile from "./Mobile";
import Desktop from "./Desktop";
import fetchDataForFilter from "../../../utils/pagination";
import ErrorBoundary from "../../ErrorBoundary/ErrorBoundary";

const PAGE_SIZE = 10;

class FixedIncome extends Component {
  constructor(props) {
    super(props);

    this.state = {
      isEmpty: false,
      activePage: 1
    };

    this.onChangePage = this.onChangePage.bind(this);
    this.onFilter = this.onFilter.bind(this);
  }

  async componentWillMount() {
    await this.props.getFixedIncomeTransactions();
    this.setState({ isEmpty: this.props.fixedIncomeTransactions.length === 0 });
  }

  onChangePage(activePage) {
    const { getFixedIncomeTransactions, filter } = this.props;
    this.setState({ activePage });
    fetchDataForFilter(
      getFixedIncomeTransactions,
      filter,
      activePage,
      PAGE_SIZE
    );
  }

  onFilter(...args) {
    this.setState({ activePage: 1 });
    this.props.getFixedIncomeTransactions(...args);
  }

  render() {
    const { fixedIncomeTransactions, loading, totalCount, error } = this.props;
    const { isEmpty, activePage } = this.state;
    return (
      <ErrorBoundary errorStatus={error}>
        <div data-test="Transactions_Fixed_Income">
          <Hide above="md">
            <Mobile
              totalCount={totalCount}
              fixedIncomeTransactions={fixedIncomeTransactions}
              loading={loading}
              isEmpty={isEmpty}
              activePage={activePage}
              onFilter={this.onFilter}
              onChangePage={this.onChangePage}
            />
          </Hide>
          <Hide below="md">
            <Desktop
              totalCount={totalCount}
              fixedIncomeTransactions={fixedIncomeTransactions}
              loading={loading}
              isEmpty={isEmpty}
              activePage={activePage}
              onFilter={this.onFilter}
              onChangePage={this.onChangePage}
            />
          </Hide>
        </div>
      </ErrorBoundary>
    );
  }
}

FixedIncome.defaultProps = {
  fixedIncomeTransactions: [],
  totalCount: 0,
  filter: {
    from: "",
    range: {
      from: "",
      to: ""
    },
    selectedOption: "",
    currentRoute: ""
  }
};

FixedIncome.propTypes = {
  fixedIncomeTransactions: arrayOf(
    shape({
      assetName: string,
      type: string,
      date: string,
      grossValue: number,
      netValue: number,
      incomeTax: number,
      iof: number
    })
  ),
  totalCount: number,
  getFixedIncomeTransactions: func.isRequired,
  filter: shape({
    from: oneOfType([string, instanceOf(Date)]),
    range: shape({
      from: oneOfType([string, instanceOf(Date)]),
      to: oneOfType([string, instanceOf(Date)])
    }),
    selectedOption: string,
    currentRoute: string
  })
};

export default FixedIncome;
