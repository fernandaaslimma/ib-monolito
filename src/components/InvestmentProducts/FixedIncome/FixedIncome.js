import React, { Component, Fragment } from "react";
import ErrorBoundary from "../../ErrorBoundary/ErrorBoundary";
import { StepSlider } from "react-bocombbm-components";
import DefaultShimmerLoading from "../../common/DefaultShimmerLoading";
import ListStep from "./ListStep";
import DetailsStep from "./DetailsStep";
import FormStep from "./FormStep";
import ConfirmationStep from "./ConfirmationStep";
import ResumeStep from "./ResumeStep";
import { isPtBR, translate } from "../../../utils/i18n";

import { InstanceContext } from "./fixedIncomeContext";
import {
  ACTION_TYPE_APPROVE_INVESTMENT,
  FIXED_INCOME,
  REGISTRATION_DATA_NOTIFICATION_TYPE,
  SUITABILITY_NOTIFICATION_TYPE
} from "../../../utils/constants";
import { MODAL_TYPES } from "../../common/Modal/Modal";
import Header from "../../common/Modal/Header/Header";
import { EFTTokenMfaWrapper } from "./styles";
import EFTToken from "../../common/EFTToken";
import { unFormatNumber } from "../../../utils/formatNumber";
import { checkDate } from "../../../utils/validations/fund";
class FixedIncome extends Component {
  constructor(props) {
    super(props);

    this.state = {
      selectedProduct: null,
      loading: true,
      filledValue: isPtBR ? "0,00" : "0.00",
      canInvest: false,
      signed: false,
      selectedAccount: {},
      typeOfFundPendencie: "",
      route: "",
      onGoingTransactions: null,
      isBottomSheetUpdate: false,
      clickedToContinue: false,
      onGoingBottomSheet: false,
      loadingButtonInvestRequest: false
    };

    this.colors = {
      moderate: "#E3F1D4",
      aggressive: "#E2DCF5",
      conservative: "#DAE6F2"
    };

    this.steps = [
      ListStep,
      DetailsStep,
      FormStep,
      ConfirmationStep,
      ResumeStep
    ];
    this.selectProduct = this.selectProduct.bind(this);
    this.changeAmmount = this.changeAmmount.bind(this);
    this.resetInvestability = this.resetInvestability.bind(this);
    this.selectAccount = this.selectAccount.bind(this);
    this.tokenModal = this.tokenModal.bind(this);
    this.createOperationFunction = this.createOperationFunction.bind(this);
    this.verifyPendencies = this.verifyPendencies.bind(this);
    this.verifyOnGoingTransactions = this.verifyOnGoingTransactions.bind(this);
    this.showFooterModal = this.showFooterModal.bind(this);
    this.resetSubscriptionPendencie = this.resetSubscriptionPendencie.bind(
      this
    );
    this.resetStateOnGoingTransaction = this.resetStateOnGoingTransaction.bind(
      this
    );
    this.handleClose = this.handleClose.bind(this);
    this.changeState = this.changeState.bind(this);
  }

  selectAccount(event) {
    const selectedAccount = this.props.accounts.find(
      account => account.accountNumber === event.target.value
    );

    this.setState({ selectedAccount });
  }

  changeAmmount(filledValue) {
    this.setState({ filledValue });
  }

  resetInvestability() {
    return this.setState({
      canInvest: false,
      signed: false,
      filledValue: isPtBR ? "0,00" : "0.00"
    });
  }

  async componentDidMount() {
    const { getInvestmentFI, getAccounts, getAvailableDateRanges } = this.props;
    await Promise.all([
      getInvestmentFI(),
      getAccounts(),
      getAvailableDateRanges(FIXED_INCOME)
    ]);

    this.setState({
      loading: false,
      selectedAccount: this.props.accounts[0]
    });
  }

  changeState(state, value) {
    this.setState({ [state]: value });
  }

  selectProduct(selectedProduct, callBack) {
    this.setState({ selectedProduct });
    callBack();
  }

  resetStateOnGoingTransaction() {
    this.setState({ onGoingTransactions: null });
  }

  resetSubscriptionPendencie() {
    this.setState({ typeOfFundPendencie: "", isBottomSheetUpdate: false });
  }

  showFooterModal(route, typeOfFundPendencie) {
    this.setState({
      typeOfFundPendencie,
      route
    });
  }

  async verifyPendencies(goToStep) {
    await this.props.getSubscriptionsPendencies();

    if (
      this.props.subscriptionsPendencies &&
      this.props.subscriptionsPendencies.length > 0
    ) {
      this.props.subscriptionsPendencies.every(item => {
        if (item.code === REGISTRATION_DATA_NOTIFICATION_TYPE) {
          this.showFooterModal(
            "/registrationData",
            REGISTRATION_DATA_NOTIFICATION_TYPE
          );
          return false;
        }
        if (item.code === SUITABILITY_NOTIFICATION_TYPE) {
          this.showFooterModal("/suitability", SUITABILITY_NOTIFICATION_TYPE);
          return false;
        }
        return true;
      });
    } else {
      goToStep(3);
    }
  }

  async verifyOnGoingTransactions(id, serverTime, goToStep) {
    this.setState({
      loadingButtonInvestRequest: true
    });
    const { getPendingTransactionsFI } = this.props;

    await getPendingTransactionsFI();

    const thisTransactionToday = this.props.pendingTransactionsFI.filter(
      transaction => {
        return (
          transaction.product.id === id &&
          checkDate(transaction.transactionDate, serverTime)
        );
      }
    );

    if (thisTransactionToday.length === 0) {
      this.createOperationFunction(goToStep);
    } else {
      this.setState({
        onGoingTransactions: thisTransactionToday,
        loadingButtonInvestRequest: false
      });
    }
  }

  async createOperationFunction(goToStep) {
    this.state.loadingButtonInvestRequest === false &&
      this.setState({ loadingButtonInvestRequest: true });
    const { createSubscriptionFI } = this.props;
    const {
      filledValue,
      selectedProduct,
      selectedAccount: { account, bankCode, branch, number }
    } = this.state;

    const body = {
      productId: selectedProduct.id,
      investmentValue: unFormatNumber(filledValue),
      account: {
        id: account,
        bankCode: bankCode,
        branch: branch,
        number: number
      }
    };
    await createSubscriptionFI(body);

    this.tokenModal(goToStep);
  }

  handleClose() {
    this.props.closeModal();
    this.setState({ signed: false });
  }

  async tokenModal(goToStep) {
    this.setState({
      loadingButtonInvestRequest: false
    });
    const { selectedProduct } = this.state;

    const {
      methodChanging,
      subscription,
      openModal,
      closeModal,
      approveSubscriptionFI,
      openToastr,
      addError
    } = this.props;

    const subscriptionId = subscription.id;

    const params = {
      actionType: ACTION_TYPE_APPROVE_INVESTMENT,
      payload: {
        id: subscriptionId,
        productName: selectedProduct.productLabel,
        ...subscription
      }
    };

    const onConfirm = async () => {
      try {
        await approveSubscriptionFI(subscriptionId);
        closeModal();
        openToastr({
          text: translate("INVESTMENTS_FIXED_INCOME_TOASTR_ACCEPTED"),
          isBelow: true,
          isTop: false,
          timeout: 3000
        });
        goToStep(5);
      } catch (error) {
        addError(error);
        closeModal();
        return { error };
      }
    };

    openModal({
      type: MODAL_TYPES.CUSTOM,
      overwriteDefaultButtons: true,
      children: () => (
        <Fragment>
          <Header
            onClickBack={
              methodChanging
                ? () => this.props.changeFactorTogle(false)
                : () => this.handleClose()
            }
            onClickClose={() => this.handleClose()}
            dataTest="MfaAddpBbmBack"
          />
          <EFTTokenMfaWrapper>
            <EFTToken
              onConfirm={onConfirm}
              onMFAConfirmation={() => onConfirm()}
              onMFAError={() => {}}
              createAuthCodeParams={params}
            />
          </EFTTokenMfaWrapper>
        </Fragment>
      )
    });
  }

  render() {
    const { loading } = this.state;
    return (
      <Fragment>
        {loading ? (
          <DefaultShimmerLoading repeat={4} innerRepeat={2} />
        ) : (
          <ErrorBoundary errorStatus={this.props.error}>
            <InstanceContext.Provider value={this}>
              <StepSlider steps={this.steps} />
            </InstanceContext.Provider>
          </ErrorBoundary>
        )}
      </Fragment>
    );
  }
}

export default FixedIncome;
