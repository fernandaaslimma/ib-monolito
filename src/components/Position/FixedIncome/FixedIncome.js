import React, { Component } from "react";
import { arrayOf, shape, number, string, func, bool } from "prop-types";

import Hide from "../../common/Hide";
import Mobile from "./Mobile";
import Desktop from "./Desktop";
import ErrorBoundary from "../../ErrorBoundary/ErrorBoundary";
import { checkDate } from "../../../utils/validations/fund";
import { redirect } from "../../../utils/redirect";
import AnimatedBottonSheet from "../../common/AnimatedBottomSheet";
import { translate } from "../../../utils/i18n";
import { BtnWrapper, Separator, StickyWrapper } from "../WithdrawLca/styles";
import Button from "react-bocombbm-components/dist/Button";
import formatNumber from "../../../utils/formatNumber";
import { BRL_CURRENCY, CREATE_TRANSACTION } from "../../../utils/constants";
import {
  ContentWrapper,
  BottomSheetWrapper,
  SheetContent,
  SheetContentContact
} from "./styles";
import { checkIfHasAccess } from "../../common/CanAccess/CanAccess";

class FixedIncome extends Component {
  constructor(props) {
    super(props);
    this.verifyOnGoingTransactions = this.verifyOnGoingTransactions.bind(this);

    this.state = {
      isEmpty: false,
      onGoingTransactions: []
    };
  }

  async componentDidMount() {
    const {
      getTotalFixedIncome,
      getFixedIncome,
      getPendingTransactionsFI,
      getTotalLca,
      getAvailableDateRanges
    } = this.props;

    if (checkIfHasAccess(this.props.userInfo, [CREATE_TRANSACTION])) {
      await Promise.all([getPendingTransactionsFI(), getTotalLca()]);
    }

    await Promise.all([
      getTotalFixedIncome(),
      getFixedIncome(),
      getAvailableDateRanges("FixedIncome")
    ]);
    this.setState({ isEmpty: this.props.fixedIncome.length === 0 });
  }

  async verifyOnGoingTransactions() {
    this.setState({
      loadingButtonWithdrawal: true
    });

    const thisTransactionToday = this.props.pendingTransactionsFI.filter(
      transaction => {
        return (
          transaction.type === "redemption" &&
          checkDate(transaction.transactionDate, this.props.serverTime)
        );
      }
    );

    if (thisTransactionToday.length === 0) {
      this.setState({ loadingButtonWithdrawal: false });

      return redirect("/investments/positions/fixed-income/withdrawal");
    } else {
      this.setState({
        onGoingTransactions: thisTransactionToday,
        onGoingTransactionValue: formatNumber(
          thisTransactionToday[0].transactionValue
        ),
        loadingButtonWithdrawal: false
      });
    }
  }

  render() {
    const { onGoingTransactions, onGoingTransactionValue } = this.state;
    const {
      fixedIncome,
      totalFixedIncome,
      loading,
      error,
      pendingTransactionsFI,
      totalLcaRaw
    } = this.props;
    const { isEmpty } = this.state;
    return (
      <ErrorBoundary errorStatus={error}>
        <div data-test="Position_Fixed_Income">
          <Hide above="md">
            <Mobile
              fixedIncome={fixedIncome}
              totalFixedIncome={totalFixedIncome}
              loading={loading}
              isEmpty={isEmpty}
              pendingTransactions={pendingTransactionsFI}
              totalLca={totalLcaRaw}
              verifyOnGoingTransactions={this.verifyOnGoingTransactions}
              userInfo={this.props.userInfo}
            />
          </Hide>
          <Hide below="md">
            <Desktop
              fixedIncome={fixedIncome}
              totalFixedIncome={totalFixedIncome}
              loading={loading}
              isEmpty={isEmpty}
              pendingTransactions={pendingTransactionsFI}
              totalLca={totalLcaRaw}
              verifyOnGoingTransactions={this.verifyOnGoingTransactions}
              userInfo={this.props.userInfo}
            />
          </Hide>
        </div>
        <AnimatedBottonSheet
          isOpen={onGoingTransactions.length > 0}
          velocity={0.3}
          head={{
            title: translate("LCA_WITHDRAW"),
            close: true
          }}
          onClickInBack={() => this.setState({ onGoingTransactions: [] })}
        >
          <BottomSheetWrapper>
            <ContentWrapper>
              <SheetContent data-test="transactionInMovement">
                {`${translate(
                  "WITHDRAWAL_LCA_ON_GOING_REDEMPTION_TEXT_1"
                )} ${BRL_CURRENCY} ${onGoingTransactionValue}. `}
              </SheetContent>

              <SheetContentContact>
                {translate("WITHDRAWAL_LCA_ON_GOING_REDEMPTION_TEXT_2")}
              </SheetContentContact>
            </ContentWrapper>
            <StickyWrapper>
              <Separator />
              <BtnWrapper>
                <Button
                  dataTest="onGoingTransactionsUnderstood"
                  onClick={() => this.setState({ onGoingTransactions: [] })}
                  spacing={{ top: "s", bottom: "s", right: "s", left: "s" }}
                >
                  {translate("WITHDRAWAL_LCA_ON_GOING_REDEMPTION_UNDERSTOOD")}
                </Button>
              </BtnWrapper>
            </StickyWrapper>
          </BottomSheetWrapper>
        </AnimatedBottonSheet>
      </ErrorBoundary>
    );
  }
}

FixedIncome.defaultProps = {
  fixedIncome: [],
  loading: false
};

FixedIncome.propTypes = {
  fixedIncome: arrayOf(
    shape({
      name: string,
      date: string,
      issuer: string,
      maturityDate: string,
      grossBalance: number,
      netBalance: number,
      portfolioShare: number,
      indexerRate: number,
      indexer: string,
      incomeTaxBalance: number,
      iofBalance: number
    })
  ),
  loading: bool,
  getFixedIncome: func.isRequired,
  getTotalFixedIncome: func.isRequired
};

export default FixedIncome;
