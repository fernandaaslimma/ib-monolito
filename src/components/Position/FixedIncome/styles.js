import styled from "styled-components";
import { gray200 } from "../../../styles/settings";
import { rem, remFontSize } from "../../../styles/tools";

export const BottomSheetWrapper = styled.div`
  max-width: ${rem(600)};
  margin: 0 auto;
`;

export const ContentWrapper = styled.div`
  display: flex;
  flex-direction: column;
  background: #f6f9fb;
  padding: ${rem(16)} ${rem(16)};
`;

export const SheetContent = styled.span`
  font-family: Lato;
  font-style: normal;
  font-weight: normal;
  font-size: ${remFontSize(14)};
  line-height: 150%;
  letter-spacing: ${rem(0.45)};
  color: ${gray200};
`;

export const SheetContentContact = styled(SheetContent)`
  margin-top: ${rem(25)};
  padding: 0;
  color: ${gray200};
`;
