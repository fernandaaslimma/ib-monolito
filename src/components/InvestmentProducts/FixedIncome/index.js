import { connect } from "redux-zero/react";
import { combineActions } from "redux-zero/utils";
import FixedIncome from "./FixedIncome";
import actions from "../../../actions/investmentProducts/FixedIncome";
import investmentProductsActions from "../../../actions/investmentProducts";
import EFTActions from "../../../actions/EFT";
import MFATokenActions from "../../../actions/mfa";
import modalActions from "../../../actions/modal";
import loginActions from "../../../actions/login";
import toastrActions from "../../../actions/toastr";

import positionActions from "../../../actions/position";
import withMaintenanceMode from "../../common/withMaintenanceMode";
import errorActions from "../../../actions/error";

const Comp = connect(
  ({
    investmentFI,
    getInvestmentFI,
    loading,
    error,
    availableDateRanges,
    serverTime,
    subscriptionsPendencies,
    authFactors,
    accounts,
    subscription,
    userInfo,
    pendingTransactionsFI,
    openToastr
  }) => ({
    investmentFI,
    getInvestmentFI,
    loading,
    error,
    availableDateRanges,
    serverTime,
    subscriptionsPendencies,
    authFactors,
    accounts,
    subscription,
    userInfo,
    pendingTransactionsFI,
    openToastr
  }),
  combineActions(
    actions,
    EFTActions,
    MFATokenActions,
    modalActions,
    loginActions,
    positionActions,
    investmentProductsActions,
    errorActions,
    toastrActions
  )
)(FixedIncome);

export default withMaintenanceMode(Comp);
