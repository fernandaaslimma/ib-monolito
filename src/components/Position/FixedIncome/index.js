import { connect } from "redux-zero/react";

import FixedIncome from "./FixedIncome";
import actions from "../../../actions/position";
import investmentProductsActions from "../../../actions/investmentProducts";
import investmentProductsActionsFI from "../../../actions/investmentProducts/FixedIncome";
import EFTActions from "../../../actions/EFT";
import loginActions from "../../../actions/login";
import withMaintenanceMode from "../../common/withMaintenanceMode";
import { combineActions } from "redux-zero/utils";

const Comp = connect(
  ({
    fixedIncome,
    totalFixedIncome,
    loading,
    getFixedIncome,
    getTotalFixedIncome,
    error,
    pendingTransactionsFI,
    totalLcaRaw,
    serverTime,
    userInfo
  }) => ({
    fixedIncome,
    totalFixedIncome,
    loading,
    getFixedIncome,
    getTotalFixedIncome,
    error,
    pendingTransactionsFI,
    totalLcaRaw,
    serverTime,
    userInfo
  }),
  combineActions(
    actions,
    investmentProductsActions,
    investmentProductsActionsFI,
    EFTActions,
    loginActions
  )
)(FixedIncome);

export default withMaintenanceMode(Comp);
