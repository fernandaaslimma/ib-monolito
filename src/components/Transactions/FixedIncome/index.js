import { connect } from "redux-zero/react";

import FixedIncome from "./FixedIncome";
import actions from "../../../actions/transactions";
import withMaintenanceMode from "../../common/withMaintenanceMode";

const Comp = connect(
  ({
    fixedIncomeTransactions,
    getFixedIncomeTransactions,
    loading,
    totalCount,
    filter,
    error
  }) => ({
    fixedIncomeTransactions,
    getFixedIncomeTransactions,
    loading,
    totalCount,
    filter,
    error
  }),
  actions
)(FixedIncome);

export default withMaintenanceMode(Comp);
