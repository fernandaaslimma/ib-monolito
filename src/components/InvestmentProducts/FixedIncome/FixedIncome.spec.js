import React from "react";
import { shallow } from "enzyme";
import FixedIncome from "./FixedIncome";
import {
  REGISTRATION_DATA_NOTIFICATION_TYPE,
  SUITABILITY_NOTIFICATION_TYPE
} from "../../../utils/constants";

const props = {
  getInvestmentFI: jest.fn(),
  getPendingTransactionsFI: jest.fn(),
  createSubscriptionFI: jest.fn(),
  getAccounts: jest.fn(),
  getAvailableDateRanges: jest.fn(),
  handleUserInputTranferDataDate: jest.fn(),
  getSubscriptionsPendencies: jest.fn(),
  createSubscription: jest.fn(() => Promise.resolve("mock")),
  UserInputTranferDataDate: jest.fn(),
  changeFactorTogle: jest.fn(),
  closeToastr: jest.fn(),
  closeModal: jest.fn(),
  openToastr: jest.fn(),
  methodChanging: true,
  openModal(config) {
    config.children();
  },
  subscriptionsPendencies: [
    {
      code: REGISTRATION_DATA_NOTIFICATION_TYPE
    },
    {
      code: SUITABILITY_NOTIFICATION_TYPE
    }
  ],
  accounts: [
    {
      account: 1,
      accountNumber: "107 2 4-3",
      blockedBalance: 7,
      availableBalance: 6,
      totalBalance: 5,
      date: "12/12",
      document: "",
      name: "",
      bankISPB: "2312321",
      verifyingDigit: 3,
      number: 4,
      branch: 2,
      bankCode: 107
    }
  ],
  subscription: {
    id: 145,
    productId: 1,
    idempotencyKey: "2b92510a-40e6-4a1e-9857-4741c5852e62",
    investmentValue: 40000.0,
    status: "AWAITING_APPROVAL"
  },
  investmentFI: [
    {
      id: 1,
      issuer: "Banco BOCOM BBM",
      issuerCnpj: "15.114.366/0002-40",
      riskProfile: "Conservative",
      riskProfileLabel: "Conservador",
      minimumSubscription: 1000,
      maximumSubscription: 500000,
      product: "LCA",
      productLabel: "LCA",
      yieldLabel: "102% DI",
      yieldIndex: "DI",
      yieldPercentual: 102,
      fixedRate: 0,
      liquidityDate: "2021-12-20",
      maturityDate: "2023-12-01",
      monthsToMaturity: 12,
      liquidityLabel: "Diária após 90 dias",
      IncomeTaxLabel: "Isento de IR"
    },
    {
      id: 2,
      issuer: "Banco BOCOM BBM",
      issuerCnpj: "15.114.366/0002-40",
      riskProfile: "Conservative",
      riskProfileLabel: "Conservador",
      minimumSubscription: 1000,
      maximunSubscription: 500000,
      product: "LCA",
      productLabel: "LCA",
      rateLabel: "105% DI",
      indexer: "DI",
      indexerRate: 105,
      fixedRate: 0,
      liquidityDate: "2023-12-01",
      maturityDate: "2023-12-01",
      monthsToMaturity: 896,
      liquidityLabel: "No vencimento",
      IncomeTaxLabel: "Isento de IR"
    }
  ],
  pendingTransactionsFI: [
    {
      id: 3122,
      type: "subscription",
      typeLabel: "Subscription",
      idempotencyKey: "06E801C5-39E7-4CA2-B7A9-29E0FF636099",
      transactionValue: 2.0,
      transactionDate: "2021-07-02",
      product: {
        id: 2,
        name: "LCA 105% DI",
        riskProfile: "Aggressive",
        riskProfileLabel: "Agressivo",
        classType: "FixedIncome",
        classTypeLabel: "Renda Fixa"
      }
    }
  ]
};

describe("InvestmentProducts component", () => {
  it("should match snapshot", () => {
    expect(shallow(<FixedIncome {...props} />)).toMatchSnapshot();
  });

  it("should change state after changeAmmount", () => {
    const instanceComponent = shallow(<FixedIncome {...props} />).instance();
    const spy = jest.spyOn(instanceComponent, "setState");

    instanceComponent.changeAmmount("1000");
    expect(spy).toHaveBeenCalledWith({ filledValue: "1000" });
  });

  it("should change state after changeAmmount", () => {
    const instanceComponent = shallow(<FixedIncome {...props} />).instance();
    const spy = jest.spyOn(instanceComponent, "setState");

    instanceComponent.resetInvestability();
    expect(spy).toHaveBeenCalledWith({
      canInvest: false,
      filledValue: "0,00",
      signed: false
    });
  });
  it("should change state after selectProduct", () => {
    const instanceComponent = shallow(<FixedIncome {...props} />).instance();
    const spy = jest.spyOn(instanceComponent, "setState");

    instanceComponent.selectProduct({}, jest.fn());
    expect(spy).toHaveBeenCalledWith({ selectedProduct: {} });
  });
  it("should change state after selectAccount", () => {
    const instanceComponent = shallow(<FixedIncome {...props} />).instance();
    const spy = jest.spyOn(instanceComponent, "setState");

    instanceComponent.selectAccount({
      target: { value: props.accounts[0].accountNumber }
    });
    expect(spy).toHaveBeenCalledWith({
      selectedAccount: {
        account: 1,
        accountNumber: "107 2 4-3",
        availableBalance: 6,
        bankCode: 107,
        bankISPB: "2312321",
        blockedBalance: 7,
        branch: 2,
        date: "12/12",
        document: "",
        name: "",
        number: 4,
        totalBalance: 5,
        verifyingDigit: 3
      }
    });
  });

  it("should change state after changeAmmount", () => {
    const instanceComponent = shallow(<FixedIncome {...props} />).instance();
    const spy = jest.spyOn(instanceComponent, "setState");

    instanceComponent.changeAmmount(1000);
    expect(spy).toHaveBeenCalledWith({ filledValue: 1000 });
  });

  it("should change state after resetStateOnGoingTransaction", () => {
    const instanceComponent = shallow(<FixedIncome {...props} />).instance();
    const spy = jest.spyOn(instanceComponent, "setState");

    instanceComponent.resetStateOnGoingTransaction();
    expect(spy).toHaveBeenCalledWith({ onGoingTransactions: null });
  });
  it("should change state after resetSubscriptionPendencie", () => {
    const instanceComponent = shallow(<FixedIncome {...props} />).instance();
    const spy = jest.spyOn(instanceComponent, "setState");

    instanceComponent.resetSubscriptionPendencie();
    expect(spy).toHaveBeenCalledWith({
      isBottomSheetUpdate: false,
      typeOfFundPendencie: ""
    });
  });

  it("should change state after showFooterModal", () => {
    const instanceComponent = shallow(<FixedIncome {...props} />).instance();
    const spy = jest.spyOn(instanceComponent, "setState");

    instanceComponent.showFooterModal("/route", "mock");
    expect(spy).toHaveBeenCalledWith({
      typeOfFundPendencie: "mock",
      route: "/route"
    });
  });

  it("should change any state after changeState", () => {
    const instanceComponent = shallow(<FixedIncome {...props} />).instance();
    const spy = jest.spyOn(instanceComponent, "setState");

    instanceComponent.changeState("isBottomSheetUpdate", true);
    expect(spy).toHaveBeenCalledWith({ isBottomSheetUpdate: true });
  });

  it("should call showFooterModal with atu cad after verifyPendencies", () => {
    const instanceComponent = shallow(<FixedIncome {...props} />).instance();

    instanceComponent.verifyPendencies();
  });

  it("should call showFooterModal with suitability after verifyPendencies", () => {
    const newProps = {
      ...props,
      subscriptionsPendencies: [
        { code: "mock" },
        { code: SUITABILITY_NOTIFICATION_TYPE }
      ]
    };
    const instanceComponent = shallow(<FixedIncome {...newProps} />).instance();

    instanceComponent.verifyPendencies();
  });

  it("should call verifyPendencies without subscriptionsPendencies", () => {
    const newProps = { ...props, subscriptionsPendencies: [] };
    const instanceComponent = shallow(<FixedIncome {...newProps} />).instance();

    instanceComponent.verifyPendencies();
  });

  it("should call verifyOnGoingTransactions and match one pending operation", () => {
    const instanceComponent = shallow(<FixedIncome {...props} />).instance();

    instanceComponent.verifyOnGoingTransactions(2, 1625220000000);
  });

  it("should call verifyOnGoingTransactions and do not match any pending operation", () => {
    const instanceComponent = shallow(<FixedIncome {...props} />).instance();

    instanceComponent.verifyOnGoingTransactions(1, 1625220000000);
  });

  it("should create subscription and invoke token modal", () => {
    const wrapperComp = shallow(<FixedIncome {...props} />);
    wrapperComp.setState({
      selectedProduct: props.investmentFI[0],
      filledValue: 10000,
      selectedAccount: {
        account: props.accounts[0].accountNumber,
        bankCode: props.accounts[0].bankCode,
        branch: props.accounts[0].branch,
        number: props.accounts[0].number
      }
    });
    wrapperComp.update();

    const wrapperInstance = wrapperComp.instance();
    wrapperInstance.createOperationFunction("mockfunc");
  });

  it("should create subscription and invoke token modal", () => {
    const instanceComponent = shallow(<FixedIncome {...props} />).instance();
    const spy = jest.spyOn(instanceComponent, "setState");

    instanceComponent.handleClose();
    expect(props.closeModal).toHaveBeenCalled();
    expect(spy).toHaveBeenCalledWith({ signed: false });
  });
});
