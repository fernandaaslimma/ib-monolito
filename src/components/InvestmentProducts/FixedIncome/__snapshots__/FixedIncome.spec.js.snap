// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InvestmentProducts component should match snapshot 1`] = `
<Fragment>
  <DefaultShimmerLoading
    innerRepeat={2}
    repeat={4}
  />
</Fragment>
`;
